packages/casting/test/fake-rpc-server.js

```diff
@@ -124,24 +131,48 @@ export const startFakeServer = (t, fakeValues, options = {}) => {
           reply(fakeStatusResult);
           break;
         }
+        case 'abci_info': {
+          const result = {
+            response: {
+              data: 'agoric',
+              version: '0.32.1',
+              last_block_height: blockHeight,
+              last_block_app_hash: '',
+            },
+          };
+          reply(result);
+          break;
+        }
         case 'abci_query': {
-          blockHeight += 2;
-          const values = fakeValues.splice(0, Math.max(1, batchSize));
-          if (values.length > 0) {
-            if (batchSize > 0) {
-              // Return a JSON stream cell.
-              const serializedValues = values.map(val =>
-                JSON.stringify(marshaller.serialize(val)),
-              );
-              responseValueBase64 = encode({
-                blockHeight: String(blockHeight - 1),
-                values: serializedValues,
-              });
-            } else {
-              // Return a single naked value.
-              responseValueBase64 = encode(marshaller.serialize(values[0]));
-            }
+          const desiredHeight = req.body.height ?? blockHeight;
+          const values = [];
+          for (
+            let batchIndex = 0;
+            batchIndex < Math.max(1, batchSize);
+            batchIndex += 1
+          ) {
+            const fakeIndex = clamp(
+              0,
+              fakeValues.length - 1,
+              desiredHeight - fakeBlocksStartHeight + batchIndex,
+            );
+            values.push(fakeValues[fakeIndex]);
+          }
+          let responseValue;
+          if (batchSize > 0) {
+            // Return a JSON stream cell.
+            const serializedValues = values.map(val =>
+              JSON.stringify(marshaller.serialize(val)),
+            );
+            responseValue = {
+              blockHeight: String(desiredHeight - 1),
+              values: serializedValues,
+            };
+          } else {
+            // Return a single naked value.
+            responseValue = JSON.stringify(marshaller.serialize(values[0]));
```

Comment:
I added JSON.stringify on this line and am suspicious. This is what made possible to unconditionally parse the JSON in `followerElementFromStreamCellValue`, but I suspect old value cells did _not_ double envelope.

I’ve made an attempt to use the new casting with an old chain, with old storage cells, and have failed to exercise scenario2 in a way that produces anything I can consume. So, I’ve got a blind spot here.

Perhaps @michaelfig can weigh in.

Reply:
I’ve pushed a follow-up that resolves my concern:

I’ve reduced the domain of `decode` from bytes to strings. We will only handle UTF-8 text cells at this time. The path forward for non-text cells is a binary stream cell encoding that distinguishes itself with a prefix that is not valid UTF-8, e.g., any control character, as a tag for the binary union.

The stream cell detector no longer uses `decode`, and explicitly uses `JSON.parse`. Then, we use `decode` for all values, regardless of whether they come from a value cell or a stream cell. This will cause a redundant parse for legacy value cells, but legacy is spelled out L E G A C Y for a reason.

I restored the shape of the test scaffold to ensure we’re properly testing legacy value cells.